daemonize.c

```diff
@@ -22,85 +22,167 @@
 #include <errno.h>
 #include <unistd.h>
 #include <signal.h>
+#include <getopt.h>
+#include <limits.h>
+
 
 #ifndef DEVNULL
 #define DEVNULL "/dev/null"
 #endif /* DEVNULL */
 
+// The permissions assigned to the redirection files
+#define RED_PERMS S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH
+
+#define OPT_OUTRED 'o'
+#define OPT_ERRRED 'e'
+#define OPT_HELP 'h'
+#define OPT_VERSION 'v'
+
 /* 
  * FIXME: This will only work in GNU CPP, because of
  * the nonstandard ## comma deletion feature
  */
 #define errx(fmt, ...)  do {						\
 		fprintf(stderr, "%s: ", name);				\
-		if (strlen(fmt)){							\
+		if (strlen(fmt)) {							\
 			fprintf(stderr, fmt, ##__VA_ARGS__);	\
 			fprintf(stderr, ": ");					\
 		}											\
 		fprintf(stderr, "%s\n", strerror(errno));	\
 		exit(EXIT_FAILURE);							\
 	} while (0)										
 
-static void print_usage();
-
 static char *name;
+static char *version = "v0.1";
 
-int main(int argc, char **argv){
-	int fd;
+static void print_usage(char **, struct option const *);
+
+
+int main(int argc, char **argv) {
+	int nullfd, outredfd, errredfd;
 	int old_errno;
 	pid_t child;
 	struct sigaction sa;
 
+	int retopt = -1;
+	struct {
+		char command[NAME_MAX];
+		char outred[NAME_MAX];
+		char errred[NAME_MAX];
+
+		char *const *arguments;
+	} pargs = {"", DEVNULL, DEVNULL, NULL};
+
+	int optindex;
+	char const *shortopts = "+ h v o: e:";
```

Comment:
One could possibly embed `OPT_OUTRED`, `OPT_ERRRED` and other abstract values into `shortopts`, but it would have been cumbersome in the code (and requires little manual editing anyway).